src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,830 @@
+
+/--
+This is an equivalent condition to be a sheaf, which is useful for the abstraction to local
+operators on elementary toposes. However this definition is defined only for sieves, not presieves.
+The equivalence between this and `is_sheaf_for` is given in `yoneda_condition_iff_sheaf_condition`.
+This version is also useful to establish that being a sheaf is preserved under isomorphism of
+presheaves.
+
+See the discussion before Equation (3) of [MM92], Chapter III, Section 4. See also C2.1.4 of
+[Elephant]. This is also a direct reformulation of https://stacks.math.columbia.edu/tag/00Z8.
+-/
+def yoneda_sheaf_condition (P : Cᵒᵖ ⥤ Type v) (S : sieve X) : Prop :=
+∀ (f : S.functor ⟶ P), ∃! g, S.functor_inclusion ≫ g = f
+
+/--
+(Implementation). This is a (primarily internal) equivalence between natural transformations
+and compatible families.
+
+Cf the discussion after Lemma 7.47.10 in https://stacks.math.columbia.edu/tag/00YW. See also
+the proof of C2.1.4 of [Elephant], and the discussion in [MM92], Chapter III, Section 4.
+-/
+def nat_trans_equiv_compatible_family :
+  (S.functor ⟶ P) ≃ {x : family_of_elements P S // x.compatible} :=
+{ to_fun := λ α,
+  begin
+    refine ⟨λ Y f hf, _, _⟩,
+    { apply α.app (op Y) ⟨_, hf⟩ },
+    { rw sieve_compatible_iff,
+      intros Y Z f g hf,
+      dsimp,
+      rw ← functor_to_types.naturality _ _ α g.op,
+      refl }
+  end,
+  inv_fun := λ t,
+  { app := λ Y f, t.1 _ f.2,
+    naturality' := λ Y Z g,
+    begin
+      ext ⟨f, hf⟩,
+      apply t.2.to_sieve_compatible _,
+    end },
+  left_inv := λ α,
+  begin
+    ext X ⟨_, _⟩,
+    refl
+  end,
+  right_inv :=
+  begin
+    rintro ⟨x, hx⟩,
+    refl,
+  end }
+
+/-- (Implementation). A lemma useful to prove `yoneda_condition_iff_sheaf_condition`. -/
+lemma extension_iff_amalgamation (x : S.functor ⟶ P) (g : yoneda.obj X ⟶ P) :
+  S.functor_inclusion ≫ g = x ↔
+  (nat_trans_equiv_compatible_family x).1.is_amalgamation (yoneda_equiv g) :=
+begin
+  change _ ↔ ∀ ⦃Y : C⦄ (f : Y ⟶ X) (h : S f), P.map f.op (yoneda_equiv g) = x.app (op Y) ⟨f, h⟩,
+  split,
+  { rintro rfl Y f hf,
+    rw yoneda_equiv_naturality,
+    dsimp,
+    simp },
+  { intro h,
+    ext Y ⟨f, hf⟩,
+    have : _ = x.app Y _ := h f hf,
+    rw yoneda_equiv_naturality at this,
+    rw ← this,
+    dsimp,
+    simp },
+end
+
+/--
+The yoneda version of the sheaf condition is equivalent to the sheaf condition.
+
+C2.1.4 of [Elephant].
+-/
+lemma yoneda_condition_iff_sheaf_condition :
+  is_sheaf_for P S ↔ yoneda_sheaf_condition P S :=
+begin
+  rw [is_sheaf_for, yoneda_sheaf_condition],
+  simp_rw [extension_iff_amalgamation],
+  rw equiv.forall_congr_left' nat_trans_equiv_compatible_family,
+  rw subtype.forall,
+  apply ball_congr,
+  intros x hx,
+  rw equiv.exists_unique_congr_left _,
+  simp,
+end
+
+/-- `P` is a sheaf for `R` iff it is separated for `R` and there exists an amalgamation. -/
+lemma separated_for_and_exists_amalgamation_iff_sheaf_for :
```

Comment:
I would prefer that the name closely follows the names occuring in the type. So this should use `is_seperated_for` and `is_amalgamation` in its name. If that's getting too long, we might want to abbreviate things, e.g., `is_amalg`.

Reply:
I've changed it a bit - I think it's useful to keep the `exists` in there though